Coding standard en Coding standard ideal bestanden gemaakt (#291)
nuno-git/druppie-fork: docs/CODING_STANDARDS.md

- **Client state**: Zustand.
- Do not over-use Context for state that could live in Zustand or React Query.

### API Client Patterns

- `api.js` currently holds all API calls. **New features should split by domain** into smaller service modules under `src/services/`, e.g. `sessionApi.js`, `approvalApi.js`.
- All API calls must handle errors and propagate meaningful messages to the UI (via Toast or ErrorBoundary).

### Linting

- **ESLint**: `npm run lint` (`.js` and `.jsx` files, `--max-warnings 0`).
- No Prettier config exists — rely on ESLint for formatting consistency.

---

## Database Standards (PostgreSQL / SQLAlchemy)

- **Engine**: PostgreSQL 15 (Alpine in Docker).
- **ORM**: SQLAlchemy models in `druppie/db/models/`.
- **Primary keys**: UUID. Use the PostgreSQL native `UUID` type. For test compatibility, the project uses a SQLite UUID shim when running on SQLite.
- **Normalization**: Relational tables are preferred. However, `Column(JSON)` is used for data with variable or open-ended schema:
  - `llm_call.request_messages` / `response_tool_calls` / `raw_request` / `raw_response`
  - `tool_call.arguments`
  - `approval.arguments`
  - `question.choices`

>   When adding a JSON column, document the expected shape in a code comment.

Reviewer: The actual count is 11 JSON columns across `db/models/`, not 7. Four are missing. Notably `question.agent_state` looks like a stable schema candidate for normalization.

```suggestion
- **Normalization**: Relational tables are preferred. However, `Column(JSON)` is used for data with variable or open-ended schema:
  - `llm_call.request_messages` / `response_tool_calls` / `raw_request` / `raw_response` / `tools_provided`
  - `tool_call.arguments`
  - `approval.arguments`
  - `question.choices` / `choices_english` / `selected_indices` / `agent_state`
  When adding a JSON column, document the expected shape in a code comment. Note: `question.agent_state` is a candidate for normalization (stable, queryable schema).
```

### Validation at Boundaries

- **Repository boundary** — Convert SQLAlchemy models to Pydantic models using `model_validate()` or explicit factory methods.
- **API boundary** — Use FastAPI `response_model` on routes returning core domain data. Evaluation/analytics routes currently do not use `response_model`.
- **External data boundary** — Validate all incoming JSON/YAML from external systems (LLM responses, MCP tools, WebSocket payloads) with Pydantic before processing.

### Type Safety Rules

1. **No `typing.cast` to silence the type checker** — Fix the underlying type issue.
2. **`**kwargs`** — Prefer explicit named parameters. One existing exception: `evaluation_repository.update_batch_run(self, batch_id: str, **kwargs)`.
3. **`# type: ignore` / `# noqa`** — Acceptable for known idioms (e.g. SQLAlchemy boolean comparison matching) and template unused imports. Do not add them for new code without justification.

---

## Security Standards

### Authentication

- **Sensitive routes** (sessions, chat, approvals, deployments) validate JWT tokens via FastAPI dependency injection (`deps.py`, `get_current_user`).

> - **Read-only / internal analytics routes** (evaluations, analytics, workspace files) do not require auth. This is current practice but should be revisited if those endpoints expose project-scoped data.

Reviewer: `routes/workspace.py` actually **does** have auth (`get_current_user` / `RequireRole`). It shouldn't be grouped with the unauthenticated analytics routes.

```suggestion
- **Read-only / internal analytics routes** (evaluations, benchmarks) do not require auth. This is current practice but should be revisited if those endpoints expose project-scoped data. Note: `workspace.py` routes **are** authenticated — they were incorrectly listed here as unauthenticated.
```

Author: The first part of the change looks good. The note, I personally find unneeded as it adds nothing.
 If we put the change through then it points out something that isn't in the document anymore.

Author: The merging of the files removed this point.

| DB tests | In-memory SQLite with UUID shim |
| Command | `cd druppie && pytest` |

**Coverage reality**: `agent_runtime/` has good test coverage. Most `services/` modules (approval, session, job, revert, etc.) lack dedicated `test_*.py` files. New services should include tests.

### Frontend E2E Tests (Playwright)

| Item | Standard |
|------|----------|
| Location | `frontend/tests/e2e/` |
| Naming | `*.spec.js` |
| Config | `frontend/playwright.config.js` |
| Run | `npm run test:e2e` |
| Fixed-delay waits | **NEVER use `page.waitForTimeout(N)`** — it pauses execution for a fixed duration regardless of page state, making tests slow and flaky. Use `waitForSelector`, `waitForResponse`, or `waitForURL` (waits until condition is met, then proceeds immediately). `waitForTimeout` exists in some existing tests but should not be added to new ones. **Test-level timeouts** (e.g. Playwright config `timeout`, `test.setTimeout()`) are safety rails and are required — they are not the same thing. |
| Workers | `workers: 1`, `fullyParallel: false` |

### Evaluation Tests

- Custom framework in `druppie/testing/` (NOT pytest):

>   - `testing/tools/` — ~80 YAML tool test definitions.

Reviewer: Same — hardcoded counts rot.

```suggestion
  - `testing/tools/` — YAML tool test definitions.
```

| Database | `postgres:15-alpine` |
| Keycloak | `quay.io/keycloak/keycloak:24.0` |
| Gitea | `gitea/gitea:1.21` |

**Rule**: Pin external image tags explicitly. Do NOT use `latest`.

**Exception**: Custom build images (e.g. sandbox) use `:latest` via environment variable fallback in `docker-compose.yml` because the image is rebuilt locally during development.

### Dockerfile Patterns

- **Backend**: Use pip cache mount for faster rebuilds:
  ```dockerfile
  RUN --mount=type=cache,target=/root/.cache/pip pip install -r requirements.txt
  ```
- **Frontend**: Multi-stage build (builder stage → runtime stage with `serve`).
- **Dev frontend**: `Dockerfile.dev` runs `npm install && npm run dev`.

### Environment Variables

- Naming: `UPPER_SNAKE_CASE` with domain prefixes:
  - `DRUPPIE_*` — application
  - `KEYCLOAK_*` — Keycloak
  - `GITEA_*` — Gitea
  - `MCP_*` — MCP servers
  - `ZAI_*`, `DEEPINFRA_*`, etc. — LLM providers
- Public URLs use `_PUBLIC_URL` suffix.
- Ports use `_PORT` suffix.
- `.env` is gitignored. `.env.example` is the template.

---

## CI/CD & Git Standards

### GitHub Actions

- `.github/workflows/build-and-deploy.yml` — Self-hosted ARC runner inside K3s:

>   - Builds 12 Docker images with BuildKit.

Reviewer: Hardcoded counts rot. Strip it.

```suggestion
  - Builds Docker images with BuildKit.
```
